lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,244 @@
+'use strict';
+
+const { SafeMap, Symbol, StringPrototypeStartsWith, SymbolAsyncIterator, Promise } = primordials;
+
+const {
+  ERR_FEATURE_UNAVAILABLE_ON_PLATFORM,
+  codes: { ERR_INVALID_ARG_TYPE, ERR_INVALID_ARG_VALUE },
+} = require('internal/errors');
+const { kEmptyObject } = require('internal/util');
+const { validateString, validateObject, validateBoolean, validateAbortSignal } = require('internal/validators');
+const { EventEmitter } = require('events');
+const path = require('path');
+
+const kFSWatchStart = Symbol('kFSWatchStart');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { readdir } = lazyLoadFsPromises();
+
+  const filenames = await readdir(dir, { withFileTypes: true });
+
+  for await (const file of filenames) {
+    const f = path.join(dir, file.name);
+
+    files.set(f, file);
+
+    if (file.isDirectory()) {
+      await traverse(f, files);
+    }
+  }
+
+  return files;
+}
+
+class FSWatcher extends EventEmitter {
+  #options = null;
+  #closed = false;
+  #files = new SafeMap();
+  #rootPath = path.resolve();
+
+  constructor(options = kEmptyObject) {
+    super();
+
+    validateObject(options, 'options');
+
+    if (options.persistent !== undefined) {
+      validateBoolean(options.persistent, 'options.persistent');
+    }
+
+    if (options.recursive !== undefined) {
+      validateBoolean(options.recursive, 'options.recursive');
+    }
+
+    if (options.signal !== undefined) {
+      validateAbortSignal(options.signal, 'options.signal');
+    }
+
+    if (options.encoding !== undefined) {
+      if (typeof options.encoding !== 'string') {
+        throw new ERR_INVALID_ARG_VALUE('encoding', 'options.encoding');
+      }
```

Comment:
```suggestion
      validateString(options.encoding, 'options.encoding');
```

Reply:
validateString throws `ERR_INVALID_ARG_TYPE`, but in order to make it 100% compatible with the other fs.watch implementations, it should throw `ERR_INVALID_ARG_VALUE`

Reply:
Ah right, I think adding a comment explaining that would be helpful in this case.

Reply:
Thanks. I just did!